app/script/auth/module/action/AuthAction.js

```diff
@@ -69,6 +69,50 @@ function doLoginPlain(loginData, onBeforeLogin, onAfterLogin) {
   };
 }
 
+export function doFinalizeSSOLogin({clientType}) {
+  return function(dispatch, getState, {apiClient, core}) {
+    dispatch(AuthActionCreator.startLogin());
+    return Promise.resolve()
+      .then(() => apiClient.init(clientType))
+      .then(() => core.init())
+      .then(() => persistAuthData(clientType, core, dispatch))
+      .then(() => dispatch(SelfAction.fetchSelf()))
+      .then(() => dispatch(CookieAction.setCookie(COOKIE_NAME_APP_OPENED, {appInstanceId: APP_INSTANCE_ID})))
+      .then(() => dispatch(ClientAction.doInitializeClient(clientType)))
+      .then(() => dispatch(AuthActionCreator.successfulLogin()))
+      .catch(error => {
+        if (error.label === BackendError.LABEL.NEW_CLIENT || error.label === BackendError.LABEL.TOO_MANY_CLIENTS) {
+          dispatch(AuthActionCreator.successfulLogin());
+        } else {
+          dispatch(AuthActionCreator.failedLogin(error));
+        }
+        throw error;
+      });
+  };
+}
+
+export function validateSSOCode(code) {
+  return function(dispatch, getState, {apiClient}) {
+    return apiClient.auth.api.headInitiateLogin(code).catch(error => {
+      let mappedError = new BackendError({code: 500, label: BackendError.SSO_ERRORS.SSO_GENERIC_ERROR});
+      if (error && error.response) {
+        switch (error.response.status) {
+          case 404: {
+            mappedError = new BackendError({code: 404, label: BackendError.SSO_ERRORS.SSO_NOT_FOUND});
+            break;
+          }
+          case 500: {
```

Comment:
Shouldn't we check for a range of errors here? Everything >= 500 is a server error.